modules/core/src/main/java/io/fluo/core/oracle/OracleClient.java

```diff
@@ -212,13 +230,28 @@ private synchronized void connect() throws IOException, KeeperException, Interru
      * Atomically closes current connection and connects to the current leader
      */
     private synchronized void reconnect() throws InterruptedException, TTransportException, KeeperException, IOException {
-      close();
+      if(transport.isOpen())
+        transport.close();
       connect();
     }
 
-    private void close() {
-      if(transport.isOpen())
+    private synchronized void close() {
+      if(transport != null && transport.isOpen())
         transport.close();
+      try {
+        if(pathChildrenCache != null)
+          pathChildrenCache.close();
```

Comment:
After we get this merged in, I can look into opening a Fluo bug and upstream curator bug.